File: backend/ticket/src/main/java/com/group4/ticket/services/TicketService.java
package com.group4.ticket.services;

import java.time.LocalDate;
import java.util.List;

import com.group4.ticket.data.model.Ticket;
import com.group4.ticket.dto.TicketDTO;
import com.group4.ticket.dto.TicketSaveDTO;

import org.apache.poi.ss.usermodel.Workbook;
import org.springframework.data.domain.Page;

import jakarta.validation.Valid;

public interface TicketService {

    // [Admin] Show list of tickets (for admin, show all)
    Page<Ticket> getAllTickets(Integer minPrice, Integer maxPrice, String sortPrice, LocalDate startDate, LocalDate endDate, int page, int size);

    // [Customer] Show list of tickets (for user history)
    Page<Ticket> getAllTicketsByUserID(String userID, Integer minPrice, Integer maxPrice, LocalDate startDate, LocalDate endDate, int page, int size);

    // [Customer] Show ticket by ID
    TicketDTO getTicketById(String id);

    // [Customer] Create new ticket
    TicketDTO createTicket(@Valid TicketSaveDTO ticketSaveDTO);

    // [-] Add payment evidence
    TicketDTO editPayment(String id, String paymentID);

    // [Admin] Change ticket status (used, unused)
    TicketDTO updateTicketStatus(String id, String status);

    // [Admin] Export list of tickets to Excel (for admin recap)
    // [Customer] Export list of tickets to Excel (for user history)
    Workbook exportTicketToExcel(List<Ticket> listOfTickets);

    // [Customer] Reschedule ticket
    TicketDTO rescheduleTicket(String id, LocalDate startDate, LocalDate endDate);
}

File: frontend/src/app/services/ticket/ticket.service.ts
import { Injectable } from '@angular/core';
import { environment } from '../../../environment/environment.prod';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Ticket, TicketDTO, TicketSaveDTO } from '../../models/ticket.model';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class TicketService {
  private apiUrl = `${environment.apiUrl}/ticket`;

  constructor(private http: HttpClient) { }

  // Create ticket
  createTicket(ticketSaveDTO: TicketSaveDTO): Observable<TicketDTO> {
    return this.http.post<TicketDTO>(this.apiUrl, ticketSaveDTO);
  }

  getAllTickets(
    minPrice?: number,
    maxPrice?: number,
    sortPrice?: string,
    sortStatus?: string,
    startDate?: string,
    endDate?: string,
    page: number = 0,
    size: number = 10
  ): Observable<{ tickets: Ticket[], currentPage: number, totalItems: number, totalPages: number }> {
    let params = new HttpParams();
    if (minPrice !== undefined) params = params.append('minPrice', minPrice.toString());
    if (maxPrice !== undefined) params = params.append('maxPrice', maxPrice.toString());
    if (sortPrice) params = params.append('sortPrice', sortPrice);
    if (sortStatus) params = params.append('sortStatus', sortStatus);
    if (startDate) params = params.append('startDate', startDate);
    if (endDate) params = params.append('endDate', endDate);
    params = params.append('page', page.toString());
    params = params.append('size', size.toString());

    return this.http.get<{ tickets: Ticket[], currentPage: number, totalItems: number, totalPages: number }>(this.apiUrl, { params });
  }

  getAllTicketsByUserID(
    minPrice?: number,
    maxPrice?: number,
    startDate?: string,
    endDate?: string,
    page: number = 0,
    size: number = 10
  ): Observable<{ tickets: Ticket[], currentPage: number, totalItems: number, totalPages: number }> {
    let params = new HttpParams();
    if (minPrice !== undefined) params = params.append('minPrice', minPrice.toString());
    if (maxPrice !== undefined) params = params.append('maxPrice', maxPrice.toString());
    if (startDate) params = params.append('startDate', startDate);
    if (endDate) params = params.append('endDate', endDate);
    params = params.append('page', page.toString());
    params = params.append('size', size.toString());

    return this.http.get<{ tickets: Ticket[], currentPage: number, totalItems: number, totalPages: number }>(`${this.apiUrl}/list`, { params });
  }

  getTicketById(id: string): Observable<TicketDTO> {
    return this.http.get<TicketDTO>(`${this.apiUrl}/${id}`);
  }

  updateTicketStatus(id: string, newStatus: string): Observable<TicketDTO> {
    return this.http.put<TicketDTO>(`${this.apiUrl}/${id}/status`, { status: newStatus });
  }

  exportTicketsToExcel(tickets: Ticket[]): Observable<Blob> {
    return this.http.post(`${this.apiUrl}/excel`, { tickets }, { responseType: 'blob' });
  }

  reduceTourQuota(paymentId: string): Observable<void> {
    const params = new HttpParams().set('paymentId', paymentId);
    return this.http.put<void>(`${this.apiUrl}/return`, null, { params });
  }
}
